File: websocket/manage_client_state/counter.py
import asyncio
import json
import logging
from websockets.asyncio.server import broadcast, serve, ServerConnection


logging.basicConfig()
USERS = set()
VALUE = 0

def users_event():
    return json.dumps({"type": "users", "count": len(USERS)})

def value_event():
    return json.dumps({"type": "cnt", "value": VALUE})

async def counter(conn: ServerConnection):
    global USERS, VALUE
    try:
        # Register user
        USERS.add(conn)
        broadcast(USERS, users_event())
        # Send current state to user
        await conn.send(value_event())
        # Manage state changes
        async for message in conn:
            event = json.loads(message)
            print(event)
            if event["key"] == "minus":
                VALUE -= 1
                broadcast(USERS, value_event())
            elif event["key"] == "plus":
                VALUE += 1
                broadcast(USERS, value_event())
            else:
                logging.error("unsupported event: %s", event)
    finally:
        # Unregister user
        USERS.remove(conn)
        broadcast(USERS, users_event())

async def main():
    async with serve(counter, "localhost", 6789):
        await asyncio.get_running_loop().create_future()  # run forever

if __name__ == "__main__":
    asyncio.run(main())

File: websocket/manage_client_state/counter.js
window.addEventListener("DOMContentLoaded", () => {
  const websocket = new WebSocket("ws://localhost:6789/");

  document.querySelector(".minus").addEventListener("click", () => {
    websocket.send(JSON.stringify({ key: "minus" }));
  });

  document.querySelector(".plus").addEventListener("click", () => {
    websocket.send(JSON.stringify({ key: "plus" }));
  });

  websocket.onmessage = ({ data }) => {
    const event = JSON.parse(data);
    switch (event.type) {
      case "cnt":
        document.querySelector(".cnt").textContent = event.value;
        break;
      case "users":
        const users = `${event.count} user${event.count == 1 ? "" : "s"}`;
        document.querySelector(".users").textContent = users;
        break;
      default:
        console.error("unsupported event", event);
    }
  };
});